File: next/dweb-browser/Helper/tests/StateTests.cs
using System;
using DwebBrowser.Helper;

namespace DwebBrowser.HelperTests;

public class StateTests
{

    [Fact]
    public async Task BaseTest()
    {
        var state1 = new State<int>(1);
        var state2 = new State<int>(1);
        var state3 = new State<int>(() => state1.Get() + state2.Get());

        Assert.Equal(2, state3.Get());
        state3.OnChange += async (value, oldValue, _) =>
        {
            Debug.WriteLine("state3:{0}", value);
        };
        state2.Set(2);
        Assert.Equal(3, state3.Get());
        Assert.Equal(1, state1.Refs.Count);
        Assert.Equal(1, state2.Refs.Count);
        Assert.Equal(2, state3.Deps.Count);

        await Task.Delay(100);
    }

    [Fact]
    public async Task ComplexTest()
    {
        var state1 = new State<bool>(false);
        var state2 = new State<int>(2);
        var state3 = new State<int>(1);
        var state4 = new State<int>(() =>
        {
            if (state1.Get())
            {
                return state2.Get();
            }
            else
            {
                return state3.Get();
            }
        });

        Assert.Equal(1, state4.Get());
        state4.OnChange += async (value, oldValue, _) =>
        {
            Debug.WriteLine("state4:{0}", value);
        };
        state1.Set(true);
        Assert.Equal(2, state4.Get());

        await Task.Delay(100);
    }

    class StatusBar
    {
        public bool Visible { get; set; }
        public bool Overlay { get; set; }
    }

    [Fact]
    public async Task OnChange_Test()
    {
        var visible = new State<bool>(false);
        var overlay = new State<bool>(false);
        //var statusBar = new StatusBar()
        //{
        //    Visible = visible.Get(),
        //    Overlay = overlay.Get()
        //};

        var statusBarState = new State<StatusBar>(() => new StatusBar()
        {
            Visible = visible.Get(),
            Overlay = overlay.Get()
        });

        statusBarState.OnChange += async (value, oldValue, _) =>
        {
            Debug.WriteLine("onChange visible value: " + value.Visible);
            Debug.WriteLine("onChange visible oldValue: " + oldValue?.Visible);
            Debug.WriteLine("onChange overlay value: " + value.Overlay);
            Debug.WriteLine("onChange overlay oldValue: " + oldValue?.Overlay);
        };

        visible.Set(true);
        var newStatusBar = statusBarState.Get();
        Debug.WriteLine("visible: {0}, overlay: {1}", newStatusBar.Visible, newStatusBar.Overlay);
    }

    struct NavigationBar
    {
        public bool Visible { get; set; }
        public bool Overlay { get; set; }
    }

    #region 用于测试是否update方法仅适用于引用类型

    [Fact]
    public async Task Update_RefType_Test()
    {
        var statusBarState = new State<StatusBar>(new StatusBar() { Visible = false, Overlay = false });
        statusBarState.Update(cache => { cache!.Visible = true; });
        Assert.True(statusBarState.Get().Visible);
    }

    [Fact]
    public async Task Update_ValueType_Test()
    {
        var navigationBarState = new State<NavigationBar>(new NavigationBar() { Visible = false, Overlay = false  });
        navigationBarState.Update(cache => { cache!.Visible = true; });
        Assert.False(navigationBarState.Get().Visible);
    }

    #endregion
}

File: android/helper/src/test/java/org/helper/StateTests.kt
package org.helper

import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.dweb_browser.helper.State
import org.junit.Assert.assertEquals
import org.junit.jupiter.api.Test

class StateTests {

  @Test
  fun simpleTest() = runBlocking {
    val state1 = State(1)
    assertEquals(1, state1.get())
    state1.set(2)
    assertEquals(2, state1.get())
  }

  @Test
  fun baseTest() = runBlocking {

    val state1 = State(1)
    val state2 = State(1)
    val state3 = State {
      state1.get() + state2.get()
    }

    assertEquals(2, state3.get())

    state3.onChange { value ->
      println("state3: $value")
    }

    state2.set(2)
    println("state2: ${state2.get()}")
    assertEquals(3, state3.get())
    assertEquals(1, state1.refs.size)
    assertEquals(1, state2.refs.size)
    assertEquals(2, state3.deps.size)

    delay(100)
  }

  @Test
  fun complexTest() = runBlocking {

    val state1 = State(false)
    val state2 = State(2)
    val state3 = State(1)

    val state4 = State(
      if (state1.get()) {
        state2.get()
      } else {
        state3.get()
      }
    )

    assertEquals(1, state4.get())

    state4.onChange { value ->
      println("state4: $value")
    }

    state1.set(true)

    assertEquals(2, state4.get())

    delay(100)
  }

  @Test
  fun onChangeTest() = runBlocking {

    val visible = State(false)
    val overlay = State(false)

    val statusBarState = State(StatusBar(visible = visible.get(), overlay = overlay.get()))

    statusBarState.onChange { value ->
      println("onChange visible value: ${value.first.visible}")
      println("onChange visible oldValue: ${value.second?.visible}")
      println("onChange overlay value: ${value.first.overlay}")
      println("onChange overlay oldValue: ${value.second?.overlay}")
    }

    visible.set(true)

    val newStatusBar = statusBarState.get()

    println("visible: ${newStatusBar.visible}, overlay: ${newStatusBar.overlay}")

  }

  // 测试update仅适用于引用类型
  @Test
  fun updateRefTypeTest() = runBlocking {

    val statusBarState = State(StatusBar(false, false))

    statusBarState.update {
      it?.visible = true
      return@update true
    }

    assert(statusBarState.get().visible)

  }

  // 测试update不适用于值类型
  @Test
  fun updateValueTypeTest() = runBlocking {

    val navigationBarState = State(NavigationBar(false, false))

    navigationBarState.update {
      it?.visible = true
      return@update true
    }

    assert(navigationBarState.get().visible)

  }

  data class StatusBar(var visible: Boolean, var overlay: Boolean)

  data class NavigationBar(var visible: Boolean, var overlay: Boolean)
}

// 其他测试函数